src/goto-symex/builtin_functions.cpp

```diff
@@ -626,6 +626,37 @@ void goto_symext::intrinsic_get_thread_state(
   symex_assign(code_assign2tc(call.ret, flag_expr), true);
 }
 
+void goto_symext::intrinsic_get_stack_size(
+  const code_function_call2t &call,
+  reachability_treet &art)
+{
+  // Get all variables
+  unsigned size = 0;
+  ns.get_context().foreach_operand([this,&size](const symbolt &s) {
```

Comment:
Here, you are getting all the variables in the program, including global variables and variables defined in other functions.

If you want to track the stack size while it is growing, simply:
1. Add a new attribute to framet, which tracks frame information.
2. Change symex_decl to update this field whenever a new variable is declared. You can also add a new option to trigger an property violation if a limit is reached.
3. You can also decrease the stack size when scoped variables go out of scope, but I'm not sure this is the right behaviour, e.g.:
```
void foo()
{
  int a, b;
  {
    int a, b;
  }
  // What is the stack size here? Four integers or two?
}
```
I'm assuming it's two or we could blow the stack by declaring variable inside for-loops, but it's better to check the standard.

Reply:
For 3 we could use the llvm bitcode as a base. All `alloca` calls are at the top of the function.

The major issue is how to map some optimizations that involves registers e.g. `mem2reg`

Reply:
I have tried to look for information about the stack size when scoped variables go out of scope, but I was unable to find anything relevant to this discussion in https://web.cs.dal.ca/~vlado/pl/C_Standard_2011-n1570.pdf and http://www2.open-std.org/JTC1/SC22/WG14/www/abq/c17_updated_proposed_fdis.pdf. I might have overlooked these documents.

Reply:
In this particular example given by @mikhailramalho

```
void foo()
{
  int a, b;
  {
    int a, b;
  }
  // What is the stack size here? Four integers or two?
}
```

LLVM produces:

$clang -emit-llvm -c stack.c
$llvm-dis stack.bc

```
; Function Attrs: noinline nounwind optnone ssp uwtable
define void @main() #0 {
  %1 = alloca i32, align 4
  %2 = alloca i32, align 4
  %3 = alloca i32, align 4
  %4 = alloca i32, align 4
  ret void
}
```

ESBMC is following the same behaviour of LLVM:

$esbmc stack.c --32 --stack-limit 127

```
Counterexample:

State 1 file stack.c line 5 function main thread 0
----------------------------------------------------
Violated property:
  file stack.c line 5 function main
  Stack limit property was violated when declaring b

VERIFICATION FAILED
```

$esbmc stack.c --32 --stack-limit 128

```
...
Generated 4 VCC(s), 0 remaining after simplification (0 assignments)
BMC program time: 0.000s

VERIFICATION SUCCESSFUL
```